pvh/cpuid.go

```diff
@@ -0,0 +1,115 @@
+package pvh
+
+import (
+	"github.com/bobuhiro11/gokvm/kvm"
+	"github.com/bobuhiro11/gokvm/probe"
+)
+
+const (
+	// When booting with PVH boot the maximum physical addressable size
+	// is a 46 bit address space even when the host supports with 5-level
+	// paging.
+	DefaultPhysBits uint32 = 46
+)
+
+// CPUID feature bits.
+const (
+	TscDeadlineTimerECXBit  uint32 = 24 // tsc deadline timer ecx bit.
+	HypervisorECXBit        uint32 = 31 // Hypervisor ecx bit.
+	MTRREDXBit              uint32 = 12 // Hypervisor ecx bit.
+	KVMFeatureAsyncPFIntBit uint8  = 14
+)
+
+// patchCPUID patches CPUIDs before vcpu generation.
+func patchCPUID(ids *kvm.CPUID) {
+	cpuidTSCPatch := kvm.CPUIDEntry2{
+		Function: 1,
+		Index:    0,
+		Flags:    0,
+		Eax:      0,
+		Ebx:      0,
+		Ecx:      TscDeadlineTimerECXBit,
+		Edx:      0,
+	}
+
+	cpuidHVPatch := kvm.CPUIDEntry2{
+		Function: 1,
+		Index:    0,
+		Flags:    0,
+		Eax:      0,
+		Ebx:      0,
+		Ecx:      HypervisorECXBit,
+		Edx:      0,
+	}
+
+	cpuidMTRRPatch := kvm.CPUIDEntry2{
+		Function: 1,
+		Index:    0,
+		Flags:    0,
+		Eax:      0,
+		Ebx:      0,
+		Ecx:      0,
+		Edx:      MTRREDXBit,
+	}
+
+	patches := []kvm.CPUIDEntry2{
+		cpuidTSCPatch,
+		cpuidHVPatch,
+		cpuidMTRRPatch,
+	}
+	for _, id := range ids.Entries {
+		for _, patch := range patches {
+			if id.Function == patch.Function && id.Index == patch.Index {
+				id.Flags |= 1 << patch.Flags
+				id.Eax |= 1 << patch.Eax
+				id.Ebx |= 1 << patch.Ebx
+				id.Ecx |= 1 << patch.Ecx
+				id.Edx |= 1 << patch.Edx
+			}
+		}
+	}
+}
```

Comment:
I thinks this func can be simpler like this.
```suggestion
// CPUID feature bits.
const (
        TscDeadlineTimerECXBit  uint32 = 1 << 24 // tsc deadline timer ecx bit.
        HypervisorECXBit        uint32 = 1 << 31 // Hypervisor ecx bit.
        MTRREDXBit              uint32 = 1 << 12 // Hypervisor ecx bit.
        KVMFeatureAsyncPFIntBit uint8  = 14
)

// patchCPUID patches CPUIDs before vcpu generation.
func patchCPUID(ids *kvm.CPUID) {
        patch := kvm.CPUIDEntry2{
                Function: 1,
                Index:    0,
                Flags:    0,
                Eax:      0,
                Ebx:      0,
                Ecx:      TscDeadlineTimerECXBit | HypervisorECXBit,
                Edx:      MTRREDXBit,
        }

        for _, id := range ids.Entries {
                if id.Function == patch.Function && id.Index == patch.Index {
                        id.Flags |= patch.Flags
                        id.Eax |= patch.Eax
                        id.Ebx |= patch.Ebx
                        id.Ecx |= patch.Ecx
                        id.Edx |= patch.Edx
                }
        }
}
```

Reply:
I will place CPUID in a seperate package and change the function to be more of a general purpose, so if anyone wants to apply a different patchset, he can use it.